main.go

```diff
@@ -0,0 +1,706 @@
+// which must verify the parameter types and arity,
+// then verify the safety we lost by piping this through EdgeX by
+// string matching the Event.Reading[].Name and JSON-unmarshal the Value string.
+//
+// Once we've reestablished these basic requirements,
+// this dispatches the content to the appropriate type-safe functions.
+func (app *inventoryApp) processEdgeXEvent(edgeXCtx *appcontext.Context, params ...interface{}) (bool, interface{}) {
+	app.sdkCtx.Store(edgeXCtx)
+
+	if len(params) != 1 {
+		if len(params) == 2 {
+			if s, ok := params[1].(string); ok && s == "" {
+				// Turns out, sometimes the "pipeline" gives a second parameter:
+				// an empty string which sometimes has type info about the first param.
+			} else {
+				err := errors.Errorf("expected a single parameter, but got a second: %T %+[1]v", params[1])
+				app.lgr.Error("Processing error.", "error", err.Error())
+				return false, err
+			}
+		} else {
+			err := errors.Errorf("expected a single parameter, but got %d", len(params))
+			app.lgr.Error("Processing error.", "error", err.Error())
+			return false, err
+		}
+	}
+
+	event, ok := params[0].(models.Event)
+	if !ok {
+		// You know what's cool in compiled languages? Type safety.
+		return false, errors.Errorf("expected an EdgeX Event, but got %T", event)
+	}
+
+	if len(event.Readings) < 1 {
+		// Is this really an error? EdgeX's Filter functions say yes.
+		return false, errors.New("event contains no Readings")
+	}
+
+	r := bytes.Buffer{}
+	decoder := json.NewDecoder(&r)
+	decoder.UseNumber()
+	decoder.DisallowUnknownFields()
+
+	for i := range event.Readings {
+		reading := &event.Readings[i] // Readings is 169 bytes. This avoid the copy.
+		switch reading.Name {
+		default:
+			app.lgr.Debug("Unknown reading.", "reading", reading.Name)
+			continue
+
+		case ResourceReaderNotification:
+			r.Reset()
+			r.WriteString(reading.Value)
+			notification := &llrp.ReaderEventNotification{}
+			if err := decoder.Decode(notification); err != nil {
+				app.lgr.Error("Failed to decode reader event notification", "error", err.Error())
+				continue
+			}
+
+			if err := app.handleReaderEvent(event.Device, notification); err != nil {
+				app.lgr.Error("Failed to handle ReaderEventNotification.",
+					"error", err.Error(), "device", event.Device)
+			}
+
+		case ResourceROAccessReport:
+			r.Reset()
+			r.WriteString(reading.Value)
+
+			report := &llrp.ROAccessReport{}
+			if err := decoder.Decode(report); err != nil {
+				app.lgr.Error("Failed to decode tag report",
+					"error", err.Error(), "device", event.Device)
+				continue
+			}
+
+			if report.TagReportData == nil {
+				app.lgr.Warn("No tag report data in report.", "device", event.Device)
+			} else {
+				app.reports <- reportData{report, inventory.NewReportInfo(reading)}
+				app.lgr.Trace("New ROAccessReport.",
+					"device", event.Device, "tags", len(report.TagReportData))
+			}
+		}
+	}
+
+	return false, nil
```

Comment:
The pipeline only consumes data and returns nil. 
Then, the taskLoop() method below calls pushEventsToCoreData(), which in turn pushes the new readings to core data directly, using sdkCtx.EventClient.Add(). 

This means that the output only goes to core-data, rather than being published to the message bus, where it could be filtered. Would it not give us more flexibility if this were to be refactored so that this processEdgeXEvent were to return events back into the pipeline, rather than for them to be pushed to core-data? @lenny-intel - what are your thoughts on this?